Update how-to-choose-identity-solution.md (#29964)
dotnet/AspNetCore.Docs: aspnetcore/security/how-to-choose-identity-solution.md

# Choose an identity management solution

> Most web apps support authentication to ensure that users are who they claim to be. A "user" might be a person or another app. Management of access ensures users are only able to see and modify the information they are authorized to see and modify. For example, an end user should not have access to the administrative section of a website. Identity management solutions are built to handle the requirements of authentication and authorization-related tasks. To learn more about identity management, see [What is identity and access management?](/azure/active-directory/fundamentals/introduction-identity-access-management)). There are many [identity management solutions for .NET web apps](xref:security/identity-management-solutions) available, each with different capabilities and requirements to use or install. This article provides guidance on how to choose the right solution.

Author: ```suggestion
Most web apps support authentication to ensure that users are who they claim to be. A *user* might be a person or another app. Management of access ensures users are only able to see and modify the information they are authorized to see and modify. For example, an end user shouldn't have access to the administrative section of a website. Identity management solutions are built to handle the requirements of authentication and authorization-related tasks. To learn more about identity management, see [What is identity and access management?](/azure/active-directory/fundamentals/introduction-identity-access-management)). There are many [identity management solutions for .NET web apps](xref:security/identity-management-solutions) available, each with different capabilities and requirements to use or install. This article provides guidance on how to choose the right solution.
```

Tokens are containers with metadata that are explicitly passed through the headers or body of HTTP requests. The main advantage of tokens over cookies is that they are not tied to a specific app or domain. Instead, tokens are usually *signed* with asymmetric cryptography. Asymmetric cryptography uses a combination of a private key known only to the signer, and a public key which everyone can know.

The token can't be tampered with due to the private key. The public key:

* Makes it possible to validate the token to ensure it hasn't been changed.
* Guarantees that it was generated by the entity holding the private key.

The main disadvantage to using tokens is that they require a token server to both issue and provide validation for tokens. The token server must be installed, configured, and maintained.

> A common reason a token server is requires is for applications that exposes web-based APIs that are consumed by other apps. For exposed web-based APIs, client UIs such as Single Page Applications (SPA), mobile clients, and desktop clients are considered to be part of the same app. SPA examples include Angular, React, and Blazor WebAssembly. If apps other than your web app or any client UIs must make a secure API call to your app, you'll likely want to use tokens. <!-- rewrite and shorten that sentence --> If you only have client UIs, ASP.NET Core Identity provides the option to acquire a token during authentication. The authentication token:

Author: ```suggestion
A common reason a token server is required is for applications that exposes web-based APIs that are consumed by other apps. For exposed web-based APIs, client UIs such as Single Page Applications (SPA), mobile clients, and desktop clients are considered to be part of the same app. SPA examples include Angular, React, and Blazor WebAssembly. If apps other than your web app or any client UIs must make a secure API call to your app, you'll likely want to use tokens. <!-- rewrite and shorten that sentence --> If you only have client UIs, ASP.NET Core Identity provides the option to acquire a token during authentication. The authentication token:
```

> A common reason a token server is required is for applications that exposes web-based APIs that are consumed by other apps. For exposed web-based APIs, client UIs such as Single Page Applications (SPA), mobile clients, and desktop clients are considered to be part of the same app. SPA examples include Angular, React, and Blazor WebAssembly. If apps other than your web app or any client UIs must make a secure API call to your app, you'll likely want to use tokens. <!-- rewrite and shorten that sentence --> If you only have client UIs, ASP.NET Core Identity provides the option to acquire a token during authentication. The authentication token:

Author: ```suggestion
A common reason a token server is required is for applications that exposes web-based APIs that are consumed by other apps. For exposed web-based APIs, client UIs such as Single Page Applications (SPA), mobile clients, and desktop clients are considered to be part of the same app. SPA examples include Angular, React, and Blazor WebAssembly. If apps other than your web app or any client UIs must make a secure API call to your app, you'll likely want to use tokens.  If you only have client UIs, ASP.NET Core Identity provides the option to acquire a token during authentication. The authentication token:
```

* Can be used by mobile and desktop clients.
* Isn't suitable for managing access from external apps.

Another reason a token server is required is for sharing sign-ins with other apps. Commonly referred to as *single sign on*, this feature enables users to:

* Sign in once with a web app's form.
* Use the resulting credentials to authenticate with other apps without having to sign-in again or choose a different password.

A token server is typically preferred to provide a secure and scalable solution for single sign on.

> For apps that don't have either of these requirements, the simplest way to quickly secure an app is to use the built-in ASP.NET Core Identity provider. Otherwise, a token server provided by a third-party identity management solution is needed. Token servers are available as:

Author: ```suggestion
For apps that don't exposes web-based APIs that are consumed by other apps and don't share logins with other apps, the simplest way to quickly secure an app is to use the built-in ASP.NET Core Identity provider. Otherwise, a token server provided by a third-party identity management solution is needed. Token servers are available as:
```

|**Customization**|✅|Self-host solutions are usually highly customizable, including code-based changes|❌|Although containerized solutions provide extensibility options, they are often more limited|❌|Cloud-based services allow customization but it is typically limited to configuration-based changes|
|**Maintenance**|❌|Installed products require a dedicated resource to ensure all security patches are applied in a timely fashion and to manage upgrades|❌|The upgrade and patch process for containers is usually lower-friction and involves simply installing the provided container image|✅|The service provider maintains their cloud-based solution, including applying needed patches and handling upgrades|
|**User credentials storage**|❌|You are responsible for data governance and handling breaches|❌|You are responsible for data governance and handling breaches|✅|The risks and regulations associated with handling user credentials are delegated to the service provider|

> |**Feature**  |**Self-host**  |**Container**  |**Cloud**  |
> |-------------:|:-:|---------------|:-:|---------------|:-:|-----------|

Reviewer: ```suggestion
|**Feature**  |**Self-host**  |**Container**  |**Cloud** |
|--------------|---------------|----------------|------------|
```

> |**Feature**  |Pro/con|**Self-host**  |Pro/con|**Container**  |Pro/con|**Cloud**  |

Author: > 

> I find this table format hard to follow. I think you could remove the Pro/Con columns and just let the text speak for itself.

I agree. How about I move the ✅ and ❌ to the `Self-host` and `Container` column.

> |**Feature**  |Pro/con|**Self-host**  |Pro/con|**Container**  |Pro/con|**Cloud**  |

Reviewer 2: Looks great with them removed. I agree.
